flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -75,10 +128,293 @@ public DefaultScheduler(
+				final CompletableFuture<LogicalSlot> slotAssigned = slotExecutionVertexAssignment.getLogicalSlotFuture();
+				checkState(slotAssigned.isDone());
+
+				FutureUtils.assertNoException(
+					slotAssigned.handle(deployOrHandleError(deploymentHandle)));
+			}
+			return null;
+		};
+	}
+
+	private static void propagateIfNonNull(final Throwable throwable) {
+		if (throwable != null) {
+			throw new CompletionException(throwable);
+		}
+	}
+
+	private BiFunction<LogicalSlot, Throwable, Void> assignResourceOrHandleError(final DeploymentHandle deploymentHandle) {
+		final ExecutionVertexVersion requiredVertexVersion = deploymentHandle.getRequiredVertexVersion();
+		final ExecutionVertexID executionVertexId = deploymentHandle.getExecutionVertexId();
+
+		return (logicalSlot, throwable) -> {
+			if (executionVertexVersioner.isModified(requiredVertexVersion)) {
+				log.debug("Refusing to assign slot to execution vertex {} because this deployment was " +
+					"superseded by another deployment", executionVertexId);
+				stopDeployment(deploymentHandle);
+				return null;
+			}
+
+			if (throwable == null) {
+				final ExecutionVertex executionVertex = getExecutionVertex(executionVertexId);
+				final boolean sendScheduleOrUpdateConsumerMessage = deploymentHandle.getDeploymentOption().sendScheduleOrUpdateConsumerMessage();
+				executionVertex
+					.getCurrentExecutionAttempt()
+					.registerProducedPartitions(logicalSlot.getTaskManagerLocation(), sendScheduleOrUpdateConsumerMessage);
+				executionVertex.tryAssignResource(logicalSlot);
+			} else {
+				handleTaskFailure(executionVertexId, throwable);
+			}
+			return null;
+		};
+	}
+
+	private BiFunction<Object, Throwable, Void> deployOrHandleError(final DeploymentHandle deploymentHandle) {
+		final ExecutionVertexVersion requiredVertexVersion = deploymentHandle.getRequiredVertexVersion();
+		final ExecutionVertexID executionVertexId = requiredVertexVersion.getExecutionVertexId();
+
+		return (ignored, throwable) -> {
+			if (executionVertexVersioner.isModified(requiredVertexVersion)) {
+				log.debug("Refusing to deploy execution vertex {} because this deployment was " +
+					"superseded by another deployment", executionVertexId);
+				stopDeployment(deploymentHandle);
+				return null;
+			}
+
+			if (throwable == null) {
+				deployTaskSafe(executionVertexId);
+			} else {
+				handleTaskFailure(executionVertexId, throwable);
+			}
+			return null;
+		};
+	}
+
+	private void stopDeployment(final DeploymentHandle deploymentHandle) {
+		cancelExecutionVertex(deploymentHandle.getExecutionVertexId());
+		deploymentHandle.getLogicalSlot().releaseSlot(null);
+	}
+
+	private void deployTaskSafe(final ExecutionVertexID executionVertexId) {
+		try {
+			final ExecutionVertex executionVertex = getExecutionVertex(executionVertexId);
+			executionVertexOperations.deploy(executionVertex);
+		} catch (Throwable e) {
+			handleTaskFailure(executionVertexId, e);
```

Comment:
If we move the error handling to the back of the chain as proposed here: https://github.com/tillrohrmann/flink/commit/546040cb7106ca371a70a6a29f28a1b08a4da2fc, then we don't need to catch the exception here.